Socket-lb: Handle connections to deleted backends

aditighag opened the pull request:
This PR addresses a limitation with socket-lb by handling stale connections to deleted service backends.

### Background

When socket-lb is enabled, traffic destined to service cluster IPs is load-balanced to service backends in the BPF cgroup hooks at the socket layer (socket `connect()` aka fast path).  When service backends are deleted, source application sockets continue to send traffic to deleted backends (particularly, for connected UDP) as there are no hooks in the slow path (e.g., socket `send()`/`receive()` calls). 

### Fix

When backends are deleted, filter host-wide sockets based on socket cookie and destination ip/port, and destroy the sockets connected to deleted backends. We use the `SOCK_DESTROY` capability in the kernel based on `NETLINK_SOCK_DIAG` infrastructure. This requires kernel to be compiled with `CONFIG_INET_DIAG_DESTROY` config.

 ### Caveats

The `SOCK_DESTROY` framework is network namespace aware, which means cilium would have to enter all network namespaces on a host, and iterate and destroy stale sockets. Cilium agent is deployed as a pod in the host netns, and we currently don't bind mount `/var/run/netns` directory on host inside the agent pod, so it doesn't see pod netns's. Options  to enable netns access: (1) run cilium agent with hostPID (security implications) so that it can scan netns (2) run this stale connections handling logic on a k8s node (similar to cilium-cni process), but this would require some kind of RPC communication channel with the agent. There are pros and cons with these options. 
For these reasons, the PR only handles stale sockets (either connected to regular backend pods or host networked backend pods) in the host netns. 

### Notes

- The long term option is to use the bpf_sock_destroy capability on kernel bpf-next/6.5+, and destroy sockets entirely in BPF using BPF iterators [1]. These changes will be done in a separate PR, and are outside of the scope of this PR.
For more details, reference the [LPC talk](https://lpc.events/event/16/contributions/1358/).

- One of the options we considered was to drop packets in the TC hooks by matching destination address. While this would have performance implications in the per packet processing, this doesn't address the issue in the host netns, as TC programs are not executed for traffic in the host netns.

TODOs

- [x] Feature detection as configs may not be enabled on all kernels
- [ ] CI test: Tested manually. New CI test will be added to the cilium-cli connectivity test suite as a follow-up. 
- [x] Document the requirements

Fixes: #12859 

```release-note
Forcefully terminate stale sockets connected to deleted service backends when socket-lb is enabled, and allow applications to re-connect to active backends.
```

[1] https://www.spinics.net/lists/bpf/msg87652.html

aditighag committed:
pkg/datapath/sockets: Infra to terminate sockets via netlink sock_diag

This commit introduces infrastructure to be able to forcefully
terminate sockets using the netlink sock_diag [1] kernel functionality.
The functionality will be used in follow-up commits to address a
limitation with socket-LB where pods can continue to connect to
deleted service backends.

[1] https://man7.org/linux/man-pages/man7/sock_diag.7.html

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
pkg/service: Refactor code to pass backends

This is a preparatory commit to pass around backend
structs instead of ids. The code will be extended in
follow-up commits to get more backend fields besides id.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
pkg/service: Handle connections to deleted service backends

Background:
When socket-lb is enabled, traffic destined to service cluster IPs
is load-balanced to service backends in the BPF cgroup hooks at the
socket layer (socket connect() aka fast path). When service backends
are deleted, application pod sockets continue to send traffic to
deleted backends (particularly, for connected UDP) as there are no
hooks in the slow path (e.g., socket send()/receive() calls) to intercept
the traffic.
We considered other options to address the limitation:
- Unreachable routes: Effective only for new connections, ICMP errors are
  ignored in established state
- Dropping traffic in Cilium TC programs: This would require adding checks
  in the per-packet LB BPF program for traffic being sent to deleted service
  backend IP:port. While per-packet LB can be disabled when socket-LB is enabled,
  this approach doesn't address the issue in the host network namespace as
  Cilium TC hooks are only executed in pod network namespaces.

Fix:
When backends are deleted, filter host-wide sockets connected to
the deleted backends. We filter load-balanced sockets based on
socket cookie and destination ip/port entries recorded in the sock
rev nat BPF map, and forcefully terminate the filtered sockets.
Sockets are forcefully terminated using the SOCK_DESTROY capability
in the kernel based on the NETLINK_SOCK_DIAG infrastructure. The functionality
is enabled via certain kernel configs, so we also check for kernel
supportability at the runtime.

Testing:

```
root@kind-worker/ # iperf3 -c 10.96.80.61  --udp -p 5010 -t 100
Connecting to host 10.96.80.61, port 5010
[  5] local 172.20.0.2 port 36895 connected to 10.96.80.61 port 5010
[ ID] Interval           Transfer     Bitrate         Total Datagrams
[  5]   0.00-1.00   sec   128 KBytes  1.05 Mbits/sec  6
[  5]   1.00-2.00   sec   128 KBytes  1.05 Mbits/sec  6
[  5]   2.00-3.00   sec   128 KBytes  1.05 Mbits/sec  6
[  5]   3.00-4.00   sec   128 KBytes  1.05 Mbits/sec  6
[  5]   4.00-5.00   sec   128 KBytes  1.05 Mbits/sec  6
[  5]   5.00-6.00   sec   128 KBytes  1.05 Mbits/sec  6
[  5]   6.00-7.00   sec   128 KBytes  1.05 Mbits/sec  6
[  5]   7.00-8.00   sec   128 KBytes  1.05 Mbits/sec  6
[  5]   8.00-9.00   sec   128 KBytes  1.05 Mbits/sec  6
[  5]   8.00-9.00   sec   128 KBytes  1.05 Mbits/sec  6
iperf3: error - unable to write to stream socket: Destination address required

level=info msg="Forcefully terminated sockets" success=1 errors="<nil>" failed=0 filter="{172.20.0.2 5010 2 17 0x3c649e0}" subsys=datapath-sockets

```

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
pkg/testutils: Mock datapath/sockets

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
pkg/service: Refactor NewService in unit test

Add a wrapper to NewService call to encapsulate
all the service fields that need to be mocked.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
pkg/service: Unit test for handling connections to deleted backends

Adds unit test cases for changes added in previous commits to
handle socket connections to deleted backends.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
pkg/maps/lbmap: Open sock rev nat map

This is a preparatory commit to open the map
in userspace in order to check if sockets are
connected to a service backend so that they
can be terminated when backends are deleted.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
pkg/maps/lbmap: Helper to check entries in sock rev nat map

Adds convenient helpers to check entries in the sock rev nat map
that's populated in the datapath for load-balancing. The helpers
are used in follow-up commits to check for sockets connected to
deleted service backends so that they can then be forcefully
terminated.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aojea commented on pkg/datapath/sockets/sockets.go:
```diff
@@ -0,0 +1,308 @@
+		if err != nil {
+			return fmt.Errorf("destroy failed: unable to get sockets with filter %v: %w", filter, err)
+		}
+		for _, sock := range sockets {
```
asking from the ignorance, do we always have to do a linear search?

aditighag replied:
Well, there is no ordering of sockets as such. The framework provides a filtering capability, so I'm looking into it so that we won't have to iterate over all sockets in the userspace.

aojea commented on pkg/datapath/sockets/sockets.go:
```diff
@@ -0,0 +1,308 @@
+				log.Debugf("Destroyed socket: %v", sock.ID)
+				success++
+			}
+		}
```
we may add a TODO for SCTP? it is GA in kubernetes and cilium has some functionality already

aditighag replied:
SCTP isn't a supported protocol with cilium's kube-proxy replacement, it's documented.

aojea commented on pkg/datapath/sockets/sockets.go:
```diff
@@ -0,0 +1,308 @@
+		for _, m := range msgs {
+			switch m.Header.Type {
+			case unix.NLMSG_DONE:
+				break loop
```
is this not the same as a simple `break` without going to the `loop`tag?

aditighag replied:
No, `break` label in Golang will exit from both the switch case and for loop in one go.

aojea commented on pkg/maps/lbmap/ipv4.go:
```diff
@@ -548,29 +548,40 @@ func (b *Backend4) GetValue() BackendValue { return b.Value }
+	RevNatIndex uint16     `align:"rev_nat_index"`
+}
+
+func (k *SockRevNat4Key) Map() *bpf.Map { return SockRevNat4Map }
```
this is a bit strange , isn't it? having a getter for a global variable

aojea commented on pkg/testutils/mockmaps/lbmap.go:
```diff
@@ -238,3 +244,19 @@ func (m *LBMockMap) UpdateSourceRanges(revNATID uint16, prevRanges []*cidr.CIDR,
+		if _, ok := m.SockRevNat4[*key]; ok {
+			return true
+		}
+	} else {
```
this has a risk if the addr `net.IP`has not been validated, because it can be nil

aditighag replied:
This is a test util code, so I'm not sure if this is required. But it doesn't hurt to do the validation I suppose.

learnitall commented on Documentation/network/kubernetes/kubeproxy-free.rst:
```diff
@@ -1617,6 +1617,13 @@ Limitations
+      namespaced that are connected to deleted backends, so that pods can be
+      load-balanced to active backends. This functionality requires these
+      kernel configs to be enabled ``CONFIG_INET_DIAG``, ``CONFIG_INET_UDP_DIAG``
+      and ``CONFIG_INET_DIAG_DESTROY``.
```
There are a couple of suggestions to help clear up the grammar in this paragraph:

```suggestion
    * When socket-LB feature is enabled, pods sending UDP traffic to services
      can continue to send traffic to a deleted service backend. Cilium agent
      handles such scenarios by forcefully terminating sockets in the host network
      namespace that are connected to deleted backends, so that pods can be
      load-balanced to active backends. This functionality requires these
      kernel configs to be enabled: ``CONFIG_INET_DIAG``, ``CONFIG_INET_UDP_DIAG``
      and ``CONFIG_INET_DIAG_DESTROY``.
```

learnitall reviewed (changes requested):
Couple of grammar nits, but otherwise docs change LGTM.

ysksuzuki commented on pkg/service/connections.go:
```diff
@@ -0,0 +1,76 @@
+	if !opSupported {
+		return
+	}
+	log.Infof("debug-aditi destroying connections to backend %v", be.L3n4Addr)
```
Should we strip `debug-aditi`?

aditighag replied:
Looks like these slipped through during rebase. I'll update them.

ysksuzuki commented on pkg/service/connections.go:
```diff
@@ -0,0 +1,76 @@
+	// records entries in the sock rev nat map that store the pod socket cookie
+	// (unique identifier in the kernel) along with the destination backend ip/port.
+	checkSockInRevNat := func(id netlink.SocketID) bool {
+		log.Infof("debug-aditi check rev nat %v %d", id, id.Cookie)
```
ditto

ysksuzuki commented on Documentation/network/kubernetes/kubeproxy-free.rst:
```diff
@@ -1617,6 +1617,13 @@ Limitations
+      handles such scenarios by forcefully terminating sockets in the host network
+      namespaced that are connected to deleted backends, so that pods can be
+      load-balanced to active backends. This functionality requires these
+      kernel configs to be enabled ``CONFIG_INET_DIAG``, ``CONFIG_INET_UDP_DIAG``
```
On ubuntu 22.04, kernel 5.19.0-50, `inet_diag` and `udp_diag` are kernel modules. It seems that those modules are loaded after destroying a UDP socket. Do you know what makes the kernel load the modules? The netlink requests that `socketDiagUDPExecutor` sends?

```
# Before
$ lsmod | head
Module                  Size  Used by
xt_TPROXY              20480  16
nf_tproxy_ipv6         20480  1 xt_TPROXY
nf_tproxy_ipv4         20480  1 xt_TPROXY
ip_set                 53248  0
xt_CT                  16384  32
xt_mark                16384  16
cls_bpf                24576  36
sch_ingress            16384  24
vxlan                 102400  0

# After destroying a socket
level=info msg="debug-aditi destroying connections to backend {10.244.1.184 {UDP 2399} 0}" subsys=service
level=info msg="debug-aditi check rev nat {38185 2399 10.244.3.92 10.244.1.184 0 [12647 0]} [12647 0]" subsys=service
level=info msg="Forcefully terminated sockets" errors="<nil>" failed=0 filter="{10.244.1.184 2399 2 17 0x25ea3c0}" subsys=datapath-sockets success=

$ lsmod | head
Module                  Size  Used by
udp_diag               16384  0
inet_diag              24576  1 udp_diag
xt_TPROXY              20480  16
nf_tproxy_ipv6         20480  1 xt_TPROXY
nf_tproxy_ipv4         20480  1 xt_TPROXY
ip_set                 53248  0
xt_CT                  16384  32
xt_mark                16384  16
cls_bpf                24576  36


# Ubuntu 22.04 kernel 5.19.0-50
$ cat /boot/config-$(uname -r) | grep INET
CONFIG_INET=y
CONFIG_INET_AH=m
CONFIG_INET_ESP=m
CONFIG_INET_ESP_OFFLOAD=m
CONFIG_INET_ESPINTCP=y
CONFIG_INET_IPCOMP=m
CONFIG_INET_TABLE_PERTURB_ORDER=16
CONFIG_INET_XFRM_TUNNEL=m
CONFIG_INET_TUNNEL=m
CONFIG_INET_DIAG=m
CONFIG_INET_TCP_DIAG=m
CONFIG_INET_UDP_DIAG=m
CONFIG_INET_RAW_DIAG=m
CONFIG_INET_DIAG_DESTROY=y
CONFIG_INET6_AH=m
CONFIG_INET6_ESP=m
CONFIG_INET6_ESP_OFFLOAD=m
CONFIG_INET6_ESPINTCP=y
CONFIG_INET6_IPCOMP=m
CONFIG_INET6_XFRM_TUNNEL=m
CONFIG_INET6_TUNNEL=m
CONFIG_INET_MPTCP_DIAG=m
CONFIG_NF_TABLES_INET=y
CONFIG_NFT_REJECT_INET=m
CONFIG_NFT_FIB_INET=m
CONFIG_NF_FLOW_TABLE_INET=m
CONFIG_INET_DCCP_DIAG=m
CONFIG_INET_SCTP_DIAG=m
CONFIG_USB_SERIAL_OMNINET=m
```

aditighag replied:
I wasn't aware of this. Thanks!
Could you confirm if the modules were loaded on the first syscall invocation, and that the agent didn't receive `EOPNOTSUPP`?

ysksuzuki replied:
`inet_diag` and `udp_diag` were automatically loaded on the first syscall invocation, and the agent didn't receive `EOPNOTSUPP` on my environment.(Ubuntu 22.04, kernel 5.19)

aditighag replied:
Thanks for confirming.

ysksuzuki commented on pkg/datapath/sockets/sockets.go:
```diff
@@ -0,0 +1,294 @@
+	return err
+}
+
+func socketDiagUDPExecutor(family uint8, receiver func(message syscall.NetlinkMessage) error) error {
```
As the `Caveats` section mentions, this implementation only lookups sockets in the host namespace. Do we encourage users to enable `bpf-lb-sock-hostns-only` if their workloads establish long-live TCP connection or connected UDP?

aditighag replied:
That's actually a good suggestion. I'll extend the docs commit to mention about this flag.

ysksuzuki commented on pkg/datapath/sockets/sockets.go:
```diff
@@ -0,0 +1,294 @@
+	// Query sockets matching the passed filter, and then destroy the filtered
+	// sockets.
+	switch protocol {
+	case unix.IPPROTO_UDP:
```
The current implementation only handles the UDP sockets, but I think the same issue occurs with long-live TCP connections. Why don't we support TCP as well?

EDIT: I found your comment here https://github.com/cilium/cilium/issues/23761#issuecomment-1479783988. I guess that the current behavior (to continue serving packets over the long-live TCP connection) is expected, although that differs from the per-packet LB. 

> Socket-lb selects a backend at the socket connect time, and as long as the backend is running, the long-lived connections should persist.

aditighag replied:
That's a fair comment.
Well, TCP has FIN/RST like signals, so if a backend is getting deleted, presumably connections are closed. At this point, we have mainly seen this issue being reported with DNS proxies using connected UDP. 
We could extend the fix for TCP connections in future if required.

ysksuzuki replied:
As Tim Hockin stated https://github.com/cilium/cilium/issues/23761#issuecomment-1478363344 , whether or not we should break the TCP connection comes down to what this pattern is used for. So, we can't decide to change the TCP behavior right now. It's out of the scope of this PR.

ysksuzuki commented on pkg/service/connections.go:
```diff
@@ -0,0 +1,76 @@
+		DestPort:  l4Addr.Port,
+		DestroyCB: checkSockInRevNat,
+	})
+	if errors.Is(err, unix.EOPNOTSUPP) {
```
We might want to log something for errors other than `unix.EOPNOTSUPP`. Otherwise, we won't know what happened here.

aditighag replied:
Sure!

ysksuzuki commented on pkg/datapath/sockets/sockets.go:
```diff
@@ -0,0 +1,294 @@
+}
+
+func (f *SocketFilter) MatchSocket(socket netlink.SocketID) bool {
+	if socket.Destination.Equal(f.DestIp) && socket.DestinationPort == f.DestPort {
```
We don't know whether a destination IP here is the one resolved from a ClusterIP or sent directly to a Pod IP. Is there a possibility of destroying sockets that we should not?

EDIT:  I tested it, but it wasn't the case. Because `MatchSocket` checks `cilium_lb[4,6]_reverse_sk` before destroying sockets. So please ignore this comment.

aditighag replied:
Yes, that's precisely why we check the map.

aditighag reviewed (commented):
@ysksuzuki Thanks for the thorough review, and giving the PR a spin! I'll address review comments, see inline responses.

aditighag committed:
docs: Document socket-LB stale connections limitation

Document the socket-LB limitation along with the
requirements of the enhancement (implemented in
previous commits) to address the limitation.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

ysksuzuki reviewed (approved):
LGTM. Thanks!

nebril reviewed (approved):
Tests LGTM.

learnitall reviewed (approved):
LGTM for docs, thank you!
